### PR TITLE
trs/coco12.cpp: add deluxe color computer

### PR DESCRIPTION
**Pull request by tlindner**
New working clones
------------------
Deluxe Color Computer (Radio Shack, 1983) [tim lindner]

Unreleased product. Boisy asked me to make this driver based on this blog post: http://latenightcoco.blogspot.com/2015/05/under-cover.html
The real ROM has not been found. I'm not sure the best way to document this.

### CODE REVIEW

**Review by npwoods — APPROVED** [review]

File: src/mame/trs/coco12.cpp [hunk 1]
```diff
@@ -708,26 +866,28 @@ ROM_END
 
 #define rom_cocoh rom_coco
 #define rom_cocoeh rom_cocoe
+#define rom_deluxecoco rom_coco2b
```
Comment (npwoods) [inline src/mame/trs/coco12.cpp:869]: The proper way to designate that we don't have the real ROM is to use:
`ROM_LOAD(..., ..., ..., NO_DUMP)`

Comment (tlindner) [reply src/mame/trs/coco12.cpp]: marked deluxecoco's ROMs as NO_DUMP

File: src/mame/trs/coco12.cpp [hunk 2]
```diff
@@ -21,6 +21,10 @@
     coco3p  1.2 2.0         1.1
     coco3h  1.2 2.0         1.1
 
+	TRS-80 Deluxe Color Computer
+
+	Prototype Color Computer with added MOS6551 and AY-3-8913
+
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:27]: Please run srcclean on this file.

File: src/mame/trs/coco12.cpp [hunk 3]
```diff
@@ -102,6 +105,13 @@ void coco12_state::ms1600_rom2(address_map &map)
...
 
+void deluxecoco_state::deluxecoco_io1(address_map &map)
+{
+	// $FF20-$FF3F
+	map(0x00, 0x03).mirror(0x1c).r(PIA1_TAG, FUNC(pia6821_device::read)).w(FUNC(coco12_state::ff20_write));
+	map(0x18, 0x19).w(m_psg, FUNC(ay8913_device::data_address_w));
+	map(0x1c, 0x1f).rw(m_acia, FUNC(mos6551_device::read), FUNC(mos6551_device::write));
+}
 
 //**************************************************************************
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:116]: Please restore the double blank line between memory maps and input ports.

File: src/mame/trs/coco12.cpp [hunk 4]
```diff
@@ -278,6 +288,109 @@ static INPUT_PORTS_START( coco )
...
+//    PA6: Ent Clr Brk Alt Ctr F1  F2 Shift
+//    PA5: 8   9   :   ;   ,   -   .   /
+//    PA4: 0   1   2   3   4   5   6   7
+//    PA3: X   Y   Z   Up  Dwn Lft Rgt Space
+//    PA2: P   Q   R   S   T   U   V   W
+//    PA1: H   I   J   K   L   M   N   O
+//    PA0: @   A   B   C   D   E   F   G
+//-------------------------------------------------
+
+static INPUT_PORTS_START( deluxecoco_keyboard )
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:307]: I can’t see this being so different to `coco_keyboard` that it warrants defining all the bits.  Use `PORT_INCLUDE(coco_keyboard)` and use `PORT_MODIFY` to change the rows that are actually different.

File: src/mame/trs/coco12.cpp [hunk 5]
```diff
@@ -562,6 +689,37 @@ void coco12_state::cocoeh(machine_config &config)
...
+	mos6551_device &acia(MOS6551(config, MOSACIA_TAG, 0));
+	acia.set_xtal(1.8432_MHz_XTAL);
+	acia.irq_handler().set(m_irqs, FUNC(input_merger_device::in_w<2>));
+	acia.txd_handler().set(ACIA_TAG, FUNC(rs232_port_device::write_txd));
+	acia.rts_handler().set(ACIA_TAG, FUNC(rs232_port_device::write_rts));
+	acia.dtr_handler().set(ACIA_TAG, FUNC(rs232_port_device::write_dtr));
+
+	rs232_port_device &rs232(RS232_PORT(config, ACIA_TAG, default_rs232_devices, nullptr));
+	rs232.set_option_device_input_defaults("null_modem", DEVICE_INPUT_DEFAULTS_NAME(acia));
+	rs232.set_option_device_input_defaults("pty", DEVICE_INPUT_DEFAULTS_NAME(acia));
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:708]: Should you set the defaults for the HLE text terminal as well?

File: src/mame/trs/coco12.cpp [hunk 6]
```diff
@@ -654,6 +812,12 @@ ROM_START(coco2)
 	ROM_LOAD("extbas11.rom",    0x0000, 0x2000, CRC(a82a6254) SHA1(ad927fb4f30746d820cb8b860ebb585e7f095dea))
 ROM_END
 
+ROM_START(deluxecoco)
+	ROM_REGION(0x8000,MAINCPU_TAG,0)
+	ROM_LOAD("bas20.rom",   0x2000, 0x2000, NO_DUMP CRC(d8f4d15e) SHA1(28b92bebe35fa4f026a084416d6ea3b1552b63d3))
+	ROM_LOAD("extbas20.rom",    0x0000, 0x2000, NO_DUMP CRC(a82a6254) SHA1(ad927fb4f30746d820cb8b860ebb585e7f095dea))
+ROM_END
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:819]: Is it even valid to have `NO_DUMP` and a CRC/SHA-1?  Usually `NO_DUMP` has no expected digest at all.

**Review by cuavas — CHANGES_REQUESTED** [review]

File: src/mame/trs/coco12.cpp (hunk 4 above)
Comment (tlindner) [reply src/mame/trs/coco12.cpp]: Yesterday, when I made this change on Windows (GCC) I got a seg fault. But now I made the same change on my Mac and it is working fine. Later today I'll test this on Windows.

File: src/mame/trs/coco12.cpp (hunk 5 above)
Comment (tlindner) [reply src/mame/trs/coco12.cpp:625]: At first I was confused that this didn't set the default for all serial options. But you ware right, Terminal is a real option as well.

File: src/mame/trs/coco12.cpp (hunk 6 above)
Comment (tlindner) [reply src/mame/trs/coco12.cpp]: Good to know. Fixed.

File: src/mame/trs/coco12.cpp [hunk 7]
```diff
@@ -278,6 +289,25 @@ static INPUT_PORTS_START( coco )
...
+	PORT_BIT(0x08, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, coco12_state, coco_state::keyboard_changed, 0) PORT_NAME("ALT") PORT_CODE(KEYCODE_LALT) PORT_CHAR(UCHAR_MAMEKEY(LALT))
+	PORT_BIT(0x10, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, coco12_state, coco_state::keyboard_changed, 0) PORT_NAME("CTRL") PORT_CODE(KEYCODE_LCONTROL) PORT_CHAR(UCHAR_MAMEKEY(LCONTROL), UCHAR_SHIFT_2)
+	PORT_BIT(0x20, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, coco12_state, coco_state::keyboard_changed, 0) PORT_CODE(KEYCODE_F1) PORT_CHAR(UCHAR_MAMEKEY(F1))
+	PORT_BIT(0x40, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, coco12_state, coco_state::keyboard_changed, 0) PORT_CODE(KEYCODE_F2) PORT_CHAR(UCHAR_MAMEKEY(F2))
+	PORT_BIT(0x80, IP_ACTIVE_LOW, IPT_KEYBOARD) PORT_CHANGED_MEMBER(DEVICE_SELF, coco12_state, coco_state::keyboard_changed, 0) PORT_NAME("SHIFT") PORT_CODE(KEYCODE_LSHIFT) PORT_CODE(KEYCODE_RSHIFT) PORT_CHAR(UCHAR_SHIFT_1)
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:306]: You only need to modify the bits that are different.  You don’t need to replace the ENTER, CLEAR, BREAK, and SHIFT inputs that are the same as they are in the `row6` port imported from the base `coco` input ports.

File: src/mame/trs/coco12.cpp [hunk 8]
```diff
@@ -644,30 +720,36 @@ ROM_END
...
-	ROM_LOAD("extbas11.rom",    0x0000, 0x2000, CRC(a82a6254) SHA1(ad927fb4f30746d820cb8b860ebb585e7f095dea))
+	ROM_LOAD("bas12.rom",    0x2000, 0x2000, CRC(54368805) SHA1(0f14dc46c647510eb0b7bd3f53e33da07907d04f))
+	ROM_LOAD("extbas11.rom", 0x0000, 0x2000, CRC(a82a6254) SHA1(ad927fb4f30746d820cb8b860ebb585e7f095dea))
+ROM_END
+
+ROM_START(deluxecoco)
+	ROM_REGION(0x8000,MAINCPU_TAG,0)
+	ROM_LOAD("bas20.rom",    0x2000, 0x2000, NO_DUMP)
+	ROM_LOAD("extbas20.rom", 0x0000, 0x2000, NO_DUMP)
 ROM_END
```
Comment (cuavas) [inline src/mame/trs/coco12.cpp:737]: Are the ROMs from the `coco2b` useful for testing this machine, even while the correct ROMs for it are undumped?  If they are, you might want to load them over the top of the `NO_DUMP` ROMs, just so people have something to play with.

Comment (tlindner) [reply src/mame/trs/coco12.cpp]: Implemented.

**Pull request merged**